### ISSUE TITLE
Next steps and tiptap 2.0

### ISSUE THREAD

**Issue by JeanLucEsser** [verbb/vizy#150]
Hey there,

I've been thinking a lot about the future of Vizy as I do think it is an amazing plugin.
So much so that I'm planning to drop Redactor for all my future clients (and even migrating some others).

I'm opening this FR because I saw tiptap is in its version 2 (beta) and I'm wondering if 1/ you're already thinking about some specific functionality of the new version and 2/ maybe already testing it along with tiptap-php to replace prosemirror.

The reason is that we already talked in previous issues as to what you were working on and I imagine that some of it may benefit or waiting on this new version. As an exemple:
- clearing only marks or nodes; 
- or switching between paragraph / headings, remember the discussion we had about toggling nodes? (#131)

I do have my list of priorities which may not be the same as yours and some of them may or may not be facilitated by tiptap 2, such as clearing/replacing on paste (#134), disabling/replacing line breaks and new paragraphs, creating custom buttons/actions (via tiptap extensions? - btw, is Vizy able to work with tiptap extensions?)

Maybe this could become a discussion about what's next, what will come out of the box with tiptap 2, what would require specific development, and getting some input from the community about priorities and such.

Feel free to close this if it doesn't make any sense to you though!

**Comment by engram-design** [comment]
There's plenty of room for growth, and we've indeed got some exciting ideas up our sleeves, but we're taking our time to develop them right.

Firstly, we're already using Tiptap 2, and have been since Vizy was released - despite it being in beta. We will be replacing the abandoned [`ueberdosis/html-to-prosemirror`](https://github.com/verbb/vizy/issues/144) at some point.

In terms of the extendability of Vizy (to write your own custom Tiptap/Prosemirror plugins) that's one of our top priorities, but we want to ensure we get the API right, as it's a layer on top of Tiptap, not to mention good documentation and examples to get you comfy with creating your own buttons and plugins for Vizy. But its [on our list](https://github.com/verbb/vizy/issues/137) to open this up.

But yes, Vizy will be able to support all extensions and plugins from Tiptap. Way down the line, we might even include the real-time [collaborative feature](https://tiptap.dev/guide/collaborative-editing) if we can get that to work with Craft.

It sounds like you're after a roadmap? There's nothing formal yet, but we're still working through Craft 4 updates (with other plugins) but we can certainly put something together

**Comment by JeanLucEsser** [comment]
Ok I see.

I'm not looking for a roadmap per se, more of a feature comparison.

For exemple, we talked about toggling between nodes in https://github.com/verbb/vizy/issues/131#issuecomment-1107058248:
> Good point, although this might be down to [Tiptap](https://tiptap.dev/) as some of this behavior is similar (casting items as Paragraph is something we can fix), I'll need to take a look.

Well, in this tiptap [exemple](https://tiptap.dev/examples/default) paragraphs and headings do toggle (apparently natively). Try going from h1 to paragraph, works perfectly. Nice thing here is that the currently selected node type is highlighted as a button which is a GREAT help. I can't see an example where they are using a formatting dropdown to see how they handle that but I guess it would be similar (unless formatting dropdown is specific to Vizy?)

See, that's the kind of information (feature parity) I'm after, more than a roadmap.

A second example, I find the clear format option to be too much, removing node AND mark formatting. But tiptap has both if the same example is to be believed (clear marks / clear nodes).

If we can know what exactly is and will be natively inherited from tiptap and what won't be, it would be of great help when thinking about writing a new FR.

Anyway, knowing that your priority is compatibility with tiptap extensions is a good thing as there may be a lot of FR that will be answered by writing a custom extension.

**Comment by engram-design** [comment]
The formatting dropdown is specific to Vizy. Tiptap is renderless, but gives you the pieces to put together a UI, and is an abstraction over ProseMirror, which itself is powerful but pretty low-level to work with.

As for the clear nodes/marks I think in practice that makes little sense to end-users (content authors), while it makes sense for us as developers who understand the terms of nodes and marks.

Typically, anything that Tiptap can do, we can. It's not so much flicking on a switch, but we need to implement UI on our end for the field, and parsing/serializing for how it renders and saves. But generally speaking, we're aiming for parity with Tiptap, if that helps.

**Comment by JeanLucEsser** [comment]
Thank you for all those details about how Tiptap/ProseMirror work, it makes a lot of sense. And that's a good thing as it leaves you a lot of leeway as to how to put together the UI, I'll keep that in mind should I file new FR.

As for the clear nodes/marks, I thought about it from an end-user perspective (we both have some UX background and sensibility from what I can see). If both options were available in Vizy, I would only use clear marks (renaming it to something user friendly) for many reasons. The first one comes from my client's mouth, clearing some bold/italic/underline in a list does remove the list all together. I think clearing nodes is not a priority for clients. It is much easier to change a node type (removing a list, changing or removing a heading...) than getting rid of a bunch of ugly formatting in a paragraph. Looking back at redactor, their clear format option only clears inline tags. Anyway, that was just an example, it's not blocking in any way, I may file a FR about it though.

Thanx again for clarifying how all the pieces come together, it really helps.
Closing this, I'll be filing individual FR if needed.

**Comment by engram-design** [comment]
Appreciate the comments, and it would be certainly possible to add all the options (just nodes, just marks, both), and the ability to re-label the button to just "Clear formatting" for ease of client understanding, and that'd be a good fit. Certainly one for the FR list!
